include/tvm/ir/diagnostic.h

```diff
@@ -0,0 +1,254 @@
+  }
+
+  static constexpr const char* _type_key = "Diagnostic";
+  TVM_DECLARE_FINAL_OBJECT_INFO(DiagnosticNode, Object);
+};
+
+class Diagnostic : public ObjectRef {
+ public:
+  TVM_DLL Diagnostic(DiagnosticLevel level, Span span, const std::string& message);
+
+  static DiagnosticBuilder Bug(Span span);
+  static DiagnosticBuilder Error(Span span);
+  static DiagnosticBuilder Warning(Span span);
+  static DiagnosticBuilder Note(Span span);
+  static DiagnosticBuilder Help(Span span);
+
+  TVM_DEFINE_OBJECT_REF_METHODS(Diagnostic, ObjectRef, DiagnosticNode);
+};
+
+/*!
+ * \brief A wrapper around std::stringstream to build a diagnostic.
+ *
+ * \code
+ *
+ * void ReportError(const Error& err);
+ *
+ * void Test(int number) {
+ *   // Use error reporter to construct an error.
+ *   ReportError(ErrorBuilder() << "This is an error number=" << number);
+ * }
+ *
+ * \endcode
+ */
+class DiagnosticBuilder {
+ public:
+  /*! \brief The level. */
+  DiagnosticLevel level;
+
+  /*! \brief The source name. */
+  SourceName source_name;
+
+  /*! \brief The span of the diagnostic. */
+  Span span;
+
+  template <typename T>
+  DiagnosticBuilder& operator<<(const T& val) {  // NOLINT(*)
+    stream_ << val;
+    return *this;
+  }
+
+  DiagnosticBuilder() : level(DiagnosticLevel::kError), source_name(), span(Span()) {}
+
+  DiagnosticBuilder(const DiagnosticBuilder& builder)
+      : level(builder.level), source_name(builder.source_name), span(builder.span) {}
+
+  DiagnosticBuilder(DiagnosticLevel level, Span span) : level(level), span(span) {}
+
+  operator Diagnostic() { return Diagnostic(this->level, this->span, this->stream_.str()); }
+
+ private:
+  std::stringstream stream_;
+  friend class Diagnostic;
+};
+
+/*!
+ * \brief A diagnostic context for recording errors against a source file.
+ */
+class DiagnosticContext;
+
+class DiagnosticContextNode : public Object {
+ public:
+  /*! \brief The source to report against. */
+  SourceMap source_map;
+
+  /*! \brief The set of diagnostics to report. */
+  Array<Diagnostic> diagnostics;
+
+  void VisitAttrs(AttrVisitor* v) {
+    v->Visit("source_map", &source_map);
+    v->Visit("diagnostics", &diagnostics);
+  }
+
+  bool SEqualReduce(const DiagnosticContextNode* other, SEqualReducer equal) const {
+    return equal(source_map, other->source_map);
+  }
+
+  static constexpr const char* _type_key = "DiagnosticContext";
+  TVM_DECLARE_FINAL_OBJECT_INFO(DiagnosticContextNode, Object);
+};
+
+class DiagnosticRenderer;
+
+class DiagnosticContext : public ObjectRef {
+ public:
+  TVM_DLL DiagnosticContext(const SourceMap& source_map);
+
+  /*! \brief Emit a diagnostic. */
+  void Emit(const Diagnostic& diagnostic);
+  /*! \brief Emit a diagnostic. */
+  void EmitFatal(const Diagnostic& diagnostic);
+  /*! \brief Render the errors and raise a DiagnosticError exception. */
+  void Render(const DiagnosticRenderer& renderer);
+
+  DiagnosticContextNode* operator->() {
+    CHECK(get() != nullptr);
```

Comment:
if we declare it as not nullable, we can remove this line.

Reply:
Did this.